src/TextInputWidget.cc

```diff
@@ -93,12 +97,50 @@ FilteredTextEdit::keyPressEvent(QKeyEvent *event)
                 }
                 break;
         }
+        case Qt::Key_V: {
+                if (event->modifiers() == Qt::ControlModifier) {
+                        insertFromMimeData(QApplication::clipboard()->mimeData());
+                }
+                break;
+        }
         default:
                 QTextEdit::keyPressEvent(event);
                 break;
         }
 }
 
+bool
+FilteredTextEdit::canInsertFromMimeData(const QMimeData *source) const
+{
+        return (source->hasImage() || source->hasText() ||
+                QTextEdit::canInsertFromMimeData(source));
+}
+
+void
+FilteredTextEdit::insertFromMimeData(const QMimeData *source)
+{
+        if (source->hasImage()) {
+                QImage img = qvariant_cast<QImage>(source->imageData());
+                pastedImagePath_ = QDir::tempPath() + '/' + "nheko_pasted_img.png";
+
+                // Save image into temporary path to be loaded later.
```

Comment:
There's no reason for this to hit disk, and doing so introduces a number of bugs. Leave the image in memory.

Reply:
The reason I write to disk is because `MatrixClient::uploadImage()` takes in a `QString` that represents a file path to an image. This means I would have to overload `uploadImage` or create a new function which accepts an image from memory. If you think that is worth doing, I will implement it.

Reply:
You should probably refactor `uploadImage` to accept data from arbitrary sources, e.g. by taking a [`QIODevice`](https://doc.qt.io/qt-5/qiodevice.html). This change is necessary for a number of reasons; most immediately, this introduces system-global state, which makes it unsafe to have multiple instances of the class even in entirely different processes.

Reply:
I'll create a new pull request with the requested changes.